### PR TITLE
Document tagging controls

### PR DESCRIPTION
**Pull request by lukavdplas**
This PR adds the ability to see and control the tags assigned to a document.

Since this is the first frontend update for tagging, it also adds a lot of the data structures for tags.

I'm marking this as a draft PR because it should only be released in combination with other tagging features in the frontend: right now, users can assign existing tags to documents but cannot create tags or do anything with them.

That said, the feature is review-ready, and I would especially like some feedback on the underlying data structures before we move on with other tagging controls.

#### Tagging controls

Screenshot of the tagging controls below.

image

image

I've tried to keep these minimal so they can be shown inline without bloating the document view. Note that they currently lack the option to make a new tag on the fly, though that could be added later. If you want to try this out, make some tags in the admin interface first.

#### Data structures

Since tags are a new model that stands mostly on its own, I've tried to implement some design principles that deviate from how data is typically handled in I-analyzer.

- Backend requests are handled by the `TagService`. The tag service uses `HttpClient` directly, rather than forwarding everything to methods of the `ApiService`. I think at this point, putting most requests in a single service is creating chaos rather than preventing it.
- Tagging logic is handled by the `TagService` (for the global list of user tags) and the `FoundDocument` model. The idea is that, for instance, the `FoundDocument` class has a `addTag` method which will trigger a backend request. This meant that that `FoundDocument` needed access to a `TagService`. Not sure about my implementation of that.
- Note the implementation of side effects. The question "what tags exist", either on the user level or the document level, occasionally needs to be refreshed when the user makes tagging actions. Notably, the component requiring a refresh may be indepedent of the component making the change. Think of a tag filter that needs to update after a user tags a document.

Note that this PR also includes #1193

### CODE REVIEW

File: frontend/src/app/tag/document-tags/document-tags.component.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+    removeTag(tag: Tag) {
+        const op = (ids: number[]) => ids.filter(id => id !== tag.id);
+        this.setTags(op);
+    }
+
+    private setTags(operation: (ids: number[]) => number[]) {
+        this.document.tags$.pipe(
+            first(),
+            map(tags => tags.map(tag => tag.id)),
```
[the remark this answers is not in the document]
Comment (JeltevanBoheemen) [reply frontend/src/app/tag/document-tags/document-tags.component.ts:38]: Not sure performs the nested map, just a suggestion.

File: frontend/src/app/tag/document-tags/document-tags.component.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    removeTag(tag: Tag) {
+        const op = (ids: number[]) => ids.filter(id => id !== tag.id);
+        this.setTags(op);
+    }
+
+    private setTags(operation: (ids: number[]) => number[]) {
+        this.document.tags$.pipe(
+            first(),
```
Comment (JeltevanBoheemen) [inline frontend/src/app/tag/document-tags/document-tags.component.ts:37]:
```suggestion
        this.document.tags$.pipe(
            takeUntil(this.destroy$),
            first(),
```
This has potential to be leaky

**Review by JeltevanBoheemen — CHANGES_REQUESTED** [review]
Looks good in general!

Bug though:
When attempting to add more than one tag:
```
tag.models.TaggedDocument.MultipleObjectsReturned: get() returned more than one TaggedDocument -- it returned 2!
```

Also not sure on the visuals. I'd like it better if the tag was in the (slightly less than `alpha:1`) color of the tag, with the cross embedded in the tag instead of being a separate button. This requires the tags to have a color though. 
Also, while I applaud the filtered tags in the select box, I think it's better if there is a single multi-select where you can search, select, and deselect all tags at once. That box could be triggered by a pencil icon in the tags component, and include a `+` button for creating new tags.

File: frontend/src/app/tag/document-tags/document-tags.component.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (lukavdplas) [reply frontend/src/app/tag/document-tags/document-tags.component.ts:37]: Oh, I assumed first was just a shorthand for take(1). Thanks!

**Comment by lukavdplas** [comment]
> I'd like it better if the tag was in the (slightly less than `alpha:1`) color of the tag, with the cross embedded in the tag instead of being a separate button. This requires the tags to have a color though.

Indeed, the current design is to get a minimal implementation going, but I did have in mind that this design would eventually work with adjustable colours.

> Also, while I applaud the filtered tags in the select box, I think it's better if there is a single multi-select where you can search, select, and deselect all tags at once. That box could be triggered by a pencil icon in the tags component, and include a `+` button for creating new tags.

I have a personal dislike of combo boxes, but perhaps that is an appropriate component to use here. I'll play around with this some more.

**Comment by lukavdplas** [comment]
This branch now includes the option to create tags on the fly, so it works fine as a standalone feature.

**Pull request merged**